clash-prelude/src/Clash/Class/Counter.hs

```diff
@@ -0,0 +1,151 @@
+{-|
+  Copyright   :  (C) 2021     , QBayLogic B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Utilities for wrapping counters consisting of multiple individual counters
+-}
+
+{-# LANGUAGE TypeFamilies #-}
+
+module Clash.Class.Counter where
+
+import Clash.Class.Num (SaturatingNum, SaturationMode(SatWrap), satSucc, satPred)
+import Clash.Sized.BitVector (BitVector)
+import Clash.Sized.Index (Index)
+import Clash.Sized.Signed (Signed)
+import Clash.Sized.Unsigned (Unsigned)
+
+import GHC.TypeLits (KnownNat, type (<=))
+
+-- | 'Counter' is a class that composes multiple counters into a single one. For
+-- example, it allows you to construct a counter from multiple smaller counters.
+-- It is similar to odometers found in olds cars, once all counters reach their
+-- maximum they reset to zero - i.e. odometer rollover. See 'countSucc' and
+-- 'countPred' for API usage examples.
+--
+-- Example use case: when driving a monitor through VGA you would like to keep
+-- track at least two counters: one counting a horizontal position, and one
+-- vertical. Perhaps a fancy VGA driver would also like to keep track of the
+-- number of drawn frames. To do so, the three counters are setup with different
+-- types. On each _round_ of the horizontal counter the vertical counter should
+-- be increased. On each _round_ of the vertical counter the frame counter should
+-- be increased. With this class you could simply use the type:
+--
+--   (FrameCount, VerticalCount, HorizontalCount)
+--
+-- and have 'countSucc' work as described.
+--
+class (Eq a, Bounded a) => Counter a where
```

Comment:
Could you drop the `Bounded a` super-class constraint so we can also have up-down counters with a non-50% duty cycle using e.g. `Counter (Either (Index 4) (Index 10))`: `Either` doesn't have a `Bounded` instance.

Reply:
Sure, following the current logic that'd mean that:

```haskell
>>> satSucc (Either (0 :: Index 4) (9 :: Index 10)) 
Either 1 0
```

I'd have to think a bit about what to change to the class. I'll get to it tomorrow.